crates/composition/src/subgraphs/selection_sets.rs

```diff
@@ -0,0 +1,120 @@
+use super::Walker;
+use crate::{strings::StringId, Subgraphs};
+use std::{
+    collections::BTreeSet,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+#[derive(Default)]
+pub(crate) struct SelectionSets(BTreeSet<Selection>);
+
+impl SelectionSets {
+    pub(crate) fn children(&self, parent: SelectionId) -> impl Iterator<Item = &Selection> {
+        self.0.range(
+            Selection {
+                parent,
+                id: SelectionId::MIN,
+                field: StringId::MIN,
+            }..Selection {
+                parent,
+                id: SelectionId::MAX,
+                field: StringId::MAX,
+            },
+        )
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct SelectionId(usize);
+
+impl SelectionId {
+    pub const MIN: SelectionId = SelectionId(usize::MIN);
+    pub const MAX: SelectionId = SelectionId(usize::MAX);
+}
+
+impl SelectionId {
+    fn next() -> SelectionId {
+        static ID_GENERATOR: AtomicUsize = AtomicUsize::new(0);
+        SelectionId(ID_GENERATOR.fetch_add(1, Ordering::Relaxed))
+    }
+}
+
+impl Subgraphs {
+    pub(crate) fn selection_set_from_str(&mut self, fields: &str) -> Result<SelectionId, String> {
+        // Cheating for now, we should import the parser from engines instead.
+        let fields = format!("{{ {fields} }}");
+        let parsed = async_graphql_parser::parse_query(fields).map_err(|err| err.to_string())?;
+
+        let async_graphql_parser::types::ExecutableDocument {
+            operations: async_graphql_parser::types::DocumentOperations::Single(operation),
+            ..
+        } = parsed
+        else {
```

Comment:
Not great formatting, but it's what rustfmt produces.

Reply:
This function is a bit long, maybe it needs to be split in two and get its own module.